## *SECTION 1 / SPRINT 1 / NOTE 2*

# Assignment


## 1. Feature Engineering

- `KT&G` 기업 데이터를 계속 사용합니다.
  - 변수 이름은 `df`로 저장하세요.

- 각 분기별 영업이익률을 직접 계산해보세요. 
  - Feature 이름은 `영업이익률2` 입니다.

In [1]:
url = 'https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g_0.csv' #파일 위치
names = ['분기', '매출액', '영업이익', '영업이익(발표기준)', '세전계속사업이익',
           '당기순이익', '당기순이익(지배)', '당기순이익(비지배)', '자산총계', '부채총계',
           '자본총계', '자본총계(지배)', '자본총계(비지배)', '자본금', '영업활동현금흐름',
           '투자활동현금흐름', '재무활동현금흐름', '영업이익률', '순이익률', 'ROE(%)',
           'ROA(%)', '부채비율', '자본유보율', 'EPS(원)', 'PER(배)'] # 데이터 내부에 열들의 명칭이 없는 경우

In [2]:
import pandas as pd

df = pd.read_csv(url, header = None)
df.columns = names

In [3]:
df.head() 

분기     매출액   영업이익 영업이익(발표기준) 세전계속사업이익  당기순이익 당기순이익(지배)  당기순이익(비지배)  \
0   19/6  12,578  3,991      3,991    4,452  3,245     3,228          17   
1   19/9  13,222  3,825      3,825    4,503  3,221     3,208          13   
2  19/12  11,982  2,523      2,523    1,794  1,183     1,198         -15   
3   20/3  11,784  3,150      3,150    4,156  2,939     2,930           9   
4   20/6  13,188  3,947      3,947    4,020  2,935     2,931           4   

      자산총계    부채총계  ... 투자활동현금흐름 재무활동현금흐름  영업이익률   순이익률 ROE(%) ROA(%)   부채비율  \
0  108,464  25,940  ...   -1,878   -4,919  31.73  25.80  12.35   9.37  31.43   
1  106,314  20,691  ...      -83     -329  28.93  24.36  12.67   9.95  24.17   
2  107,121  20,062  ...     -501     -129  21.06   9.87  12.39   9.96  23.04   
3  108,594  24,862  ...   -1,392      -86  26.73  24.94  13.05   9.79  29.69   
4  110,282  23,386  ...     -214   -5,501  29.93  22.25  12.20   9.40  26.91   

    자본유보율  EPS(원)  PER(배)  
0  803.22   2,351   13.77  
1  836.86   2,337   13.87  
2  850.82     873   12.42  
3  823.16   2,134    9.71  
4  837.40   2,135   10.46  

[5 rows x 25 columns]

### lambda 함수 

- 한마디로 함수를 한줄로 표현하게 해준다. 

``` python 
lambda x : x.replace(".","") # : 왼쪽이 함수 공식(?) 오른쪽이 반환값
```

### apply 함수 

- 파라미터에 함수 조건식을 넣으면 데이터를 하나하나씩 넣고 반환해준다. 

In [4]:
df.영업이익 = df.영업이익.apply(lambda x : x.replace(',','')) # 데이터 속 , 빼기 
df.영업이익 = df.영업이익.astype(int) #데이터 프레임 속 열 타입 바꾸기 

In [5]:
df.매출액 = df.매출액.apply(lambda x : x.replace(',','')) # 데이터 속 , 빼기 
df.매출액 = df.매출액.astype(int) #데이터 프레임 속 열 타입 바꾸기 

In [7]:
# 열추가 

df['영업이익률2'] = df['영업이익']/df['매출액']*100

- 아래 데이터를 참조하여 18/19년도에 대해서 Price to Dream Ratio (PDR)을 계산하세요.
  - 변수 이름은 `df2`, feature 이름은 `PDR`로, `df2['PDR']`을 통해 `~` 값이 확인 되어야합니다. 아래의 전제조건을 활용하세요.
  - 점유율 (일반담배 기준) / feature 이름 은 컬럼 이름
  
    - 18년 : 62%
    - 19년 : 63.5% 
  - 시장의 규모 
     - 18년 : 15조 
     - 19년 : 15조

||구분|종가|발행주식수|시가총액|시장점유율|PDR|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|0|18년|101500|137292497|13935188445500|62.0|~|
|1|19년|93800|137292497|12878036218600|63.5|~|


In [25]:
names = ['구분', '종가', '발행주식수', '시가총액', '시장점유율']

In [26]:
df2 = pd.DataFrame([['18년', 101500, 137292497, 13935188445500, 62.0], 
                   ['19년', 93800, 137292497, 12878036218600, 63.5]],
                   columns = names)
df2

구분      종가      발행주식수            시가총액  시장점유율
0  18년  101500  137292497  13935188445500   62.0
1  19년   93800  137292497  12878036218600   63.5

In [ ]:
PDR = 시가총액 / 시장 규모  * 시장점유율

In [27]:
df2['PDR'] = df2['시가총액']/15000000000000*df2['시장점유율']

In [28]:
df2

구분      종가      발행주식수            시가총액  시장점유율        PDR
0  18년  101500  137292497  13935188445500   62.0  57.598779
1  19년   93800  137292497  12878036218600   63.5  54.517020

## 2. String Manipulation

- replace가 아닌 다른 방법을 사용하여 19년도 2분기의 매출액 을 integer로 변경하세요. 결과값은 12578이 나와야합니다.
  


In [45]:
df.loc[0,'매출액'].astype(int)

12578

## 3. Apply
- 데이터셋의 분기를 제외한 모든 feature에 대해서 수치형 데이터 (`int`, `float`) 로 수정하고, `describe`를 통해 summary statistics 를 확인하세요.
- 단 결측치는 `0`으로 처리합니다.

In [65]:
df.매출액

0    12578
1    13222
2    11982
3    11784
4    13188
Name: 매출액, dtype: int32

In [73]:
fill = df.iloc[:,1:]
fill = fill.fillna(0)

for i in ['영업이익(발표기준)', '세전계속사업이익', '당기순이익', '당기순이익(지배)',
       '자산총계', '부채총계', '자본총계', '자본총계(지배)', '자본금',
       '영업활동현금흐름', '투자활동현금흐름', '재무활동현금흐름', 'EPS(원)'] :
    
    fill['{}'.format(i)] = fill['{}'.format(i)].apply(lambda x : x.replace(',',''))
    fill['{}'.format(i)] = fill['{}'.format(i)].astype(int)

In [75]:
fill.describe()

매출액         영업이익   영업이익(발표기준)     세전계속사업이익        당기순이익  \
count      5.000000     5.000000     5.000000     5.000000     5.000000   
mean   12550.800000  3487.200000  3487.200000  3785.000000  2704.600000   
std      664.979097   636.931079   636.931079  1131.079131   863.422724   
min    11784.000000  2523.000000  2523.000000  1794.000000  1183.000000   
25%    11982.000000  3150.000000  3150.000000  4020.000000  2935.000000   
50%    12578.000000  3825.000000  3825.000000  4156.000000  2939.000000   
75%    13188.000000  3947.000000  3947.000000  4452.000000  3221.000000   
max    13222.000000  3991.000000  3991.000000  4503.000000  3245.000000   

         당기순이익(지배)  당기순이익(비지배)           자산총계         부채총계         자본총계  ...  \
count     5.000000    5.000000       5.000000      5.00000      5.00000  ...   
mean   2699.000000    5.600000  108155.000000  22988.20000  85166.80000  ...   
std     851.338358   12.481987    1522.396795   2560.38419   1988.82596  ...   
min    1198.000000  -15.000000  106314.000000  20062.00000  82524.00000  ...   
25%    2930.000000    4.000000  107121.000000  20691.00000  83732.00000  ...   
50%    2931.000000    9.000000  108464.000000  23386.00000  85623.00000  ...   
75%    3208.000000   13.000000  108594.000000  24862.00000  86896.00000  ...   
max    3228.000000   17.000000  110282.000000  25940.00000  87059.00000  ...   

         재무활동현금흐름      영업이익률       순이익률     ROE(%)    ROA(%)       부채비율  \
count     5.00000   5.000000   5.000000   5.000000  5.000000   5.000000   
mean  -2192.80000  27.676000  21.444000  12.532000  9.694000  27.048000   
std    2763.51139   4.115651   6.601267   0.335738  0.290224   3.554859   
min   -5501.00000  21.060000   9.870000  12.200000  9.370000  23.040000   
25%   -4919.00000  26.730000  22.250000  12.350000  9.400000  24.170000   
50%    -329.00000  28.930000  24.360000  12.390000  9.790000  26.910000   
75%    -129.00000  29.930000  24.940000  12.670000  9.950000  29.690000   
max     -86.00000  31.730000  25.800000  13.050000  9.960000  31.430000   

           자본유보율       EPS(원)     PER(배)     영업이익률2  
count    5.00000     5.000000   5.000000   5.000000  
mean   830.29200  1966.000000  12.046000  27.675106  
std     18.01957   619.939513   1.898112   4.116711  
min    803.22000   873.000000   9.710000  21.056585  
25%    823.16000  2134.000000  10.460000  26.731161  
50%    836.86000  2135.000000  12.420000  28.929058  
75%    837.40000  2337.000000  13.770000  29.928723  
max    850.82000  2351.000000  13.870000  31.730005  

[8 rows x 25 columns]

## 🔥 도전과제



### NA Value Handling

- 19년도 4분기의 당기순이익(비지배) 부분을 Na로 대체하세요
- 이후 해당 결측치를 mean imputation 방법을 사용하여 처리하세요.



In [ ]:
# 이 곳에 답안을 작성하시길 바랍니다

### Feature Engineering

Relative Perfomance 라는 새로운 feature를 계산하세요.

이는 최근 1년치 매출액의 평균값을 기준으로 

- 10% 이상 -> `S`
- 5% 이상 -> `A`
- -5 ~ 5% -> `B`
- -5%이하 -> `C`
- 10%이하 -> `D`

라는 값을 갖는 feature입니다.

20년도 2분기에 해당하는 결과값은 `A`가 나와야합니다.

각각에 해당하는 등급이 나오기 위해서 필요한 매출액을 추가로 서술하세요.

In [ ]:
# 이 곳에 답안을 작성하시길 바랍니다